File: webapp/src/types/events.ts
import type { AgentId, AgentModelConfig } from "./agent";
import type { Artifact, ArtifactStatus, ArtifactVisibility } from "./artifact";
import type { ProjectMode, UserInputRequest } from "./project";
import type { Reference } from "./reference";
import type { WorkflowStageId } from "./workflow";

/** 05 판정(CLAUDE.md §5). 첫 줄로 정해진다. */
export type ValidationVerdict = "passed" | "conditional" | "blocked";

/**
 * 오케스트레이터 → UI 이벤트. 화면의 모든 변화는 이 이벤트로만 일어난다.
 * 백엔드(webapp/server)도 같은 모양으로 보낸다 — 타입을 바꾸면 server/app/events.py도 맞춘다.
 */
export type WorkflowEventPayload =
  | { type: "workflow.started" }
  | { type: "workflow.stage.started"; stageId: WorkflowStageId }
  | { type: "workflow.stage.completed"; stageId: WorkflowStageId }
  | { type: "workflow.warning"; stageId: WorkflowStageId; message: string }
  | { type: "workflow.completed" }
  /** 사용자가 중지를 요청함 — 현재 단계가 끝나면 멈춘다(graceful stop) */
  | { type: "workflow.stop.requested" }
  | { type: "workflow.stopped"; stageId: WorkflowStageId }
  | { type: "workflow.failed"; stageId: WorkflowStageId; reason: string }
  | { type: "project.mode.decided"; mode: ProjectMode }
  | { type: "validation.verdict"; verdict: ValidationVerdict; firstLine: string }
  | { type: "user.input.required"; request: UserInputRequest }
  | { type: "user.input.resolved"; promptId: string; answer: string }
  | { type: "agent.started"; agentId: AgentId; stageId: WorkflowStageId }
  | { type: "agent.message"; agentId: AgentId; toAgentId?: AgentId; text: string }
  | { type: "agent.activity"; agentId: AgentId; label: string }
  | { type: "agent.completed"; agentId: AgentId }
  | { type: "agent.configured"; agentId: AgentId; config: AgentModelConfig }
  /** updatedAt은 이벤트 시각(at)으로 채운다 */
  | { type: "artifact.created"; artifact: Omit<Artifact, "updatedAt"> }
  | {
      type: "artifact.updated";
      artifactId: string;
      status: ArtifactStatus;
      visibility?: ArtifactVisibility;
      summary?: string;
      /** 옛 버전으로 내려가는 것처럼 알릴 필요 없는 갱신 — 피드에 카드를 만들지 않는다 */
      silent?: boolean;
    }
  | { type: "handoff.created"; fromAgentId: AgentId; toAgentId: AgentId; artifactName?: string }
  /** addedAt은 이벤트 시각(at)으로 채운다 */
  | { type: "reference.added"; reference: Omit<Reference, "addedAt"> }
  | { type: "reference.removed"; referenceId: string }
  | { type: "user.message"; text: string };

export const EVENT_SCHEMA_VERSION = 1;

/** 전송 정보. 컴포넌트는 이 필드를 몰라도 된다(서비스·reducer만 쓴다). */
export interface EventEnvelope {
  schemaVersion: number;
  id: string;
  projectId: string;
  /** 실행 밖에서 생긴 이벤트(레퍼런스 추가 등)는 null */
  runId: string | null;
  /** 프로젝트 안에서 1부터 계속 증가. 이어 받기·중복 제거의 기준 */
  seq: number;
  at: string;
}

export type WorkflowEvent = WorkflowEventPayload & EventEnvelope;

File: webapp/server/app/events.py
"""EventStore — 이벤트 추가(seq 부여) · 다시 보내기(replay) · SSE 구독자에게 전달.

이벤트 모양은 프론트엔드 src/types/events.ts와 같다. 타입을 더하면 두 곳을 함께 고친다
(tests/test_contract_sync.py가 두 목록이 같은지 확인한다).
"""

from __future__ import annotations

import asyncio
import json
import threading
import uuid

from .db import Database, now_iso

SCHEMA_VERSION = 1

EVENT_TYPES = frozenset(
    {
        "workflow.started",
        "workflow.stage.started",
        "workflow.stage.completed",
        "workflow.warning",
        "workflow.completed",
        "workflow.stop.requested",
        "workflow.stopped",
        "workflow.failed",
        "project.mode.decided",
        "validation.verdict",
        "user.input.required",
        "user.input.resolved",
        "agent.started",
        "agent.message",
        "agent.activity",
        "agent.completed",
        "agent.configured",
        "artifact.created",
        "artifact.updated",
        "handoff.created",
        "reference.added",
        "reference.removed",
        "user.message",
    }
)


def _to_event(row) -> dict:
    return {
        **json.loads(row["payload_json"]),
        "schemaVersion": SCHEMA_VERSION,
        "id": row["id"],
        "projectId": row["project_id"],
        "runId": row["run_id"],
        "seq": row["seq"],
        "at": row["created_at"],
    }


class _Subscriber:
    def __init__(self, loop: asyncio.AbstractEventLoop):
        self.loop = loop
        self.queue: asyncio.Queue[dict] = asyncio.Queue()


class EventStore:
    """seq는 프로젝트마다 1부터 계속 증가한다. 한 번 쓴 이벤트는 바꾸지 않는다."""

    def __init__(self, db: Database):
        self._db = db
        self._subscribers: dict[str, set[_Subscriber]] = {}
        self._sub_lock = threading.Lock()

    def append(self, project_id: str, payload: dict, run_id: str | None = None) -> dict:
        event_type = payload.get("type")
        if event_type not in EVENT_TYPES:
            raise ValueError(f"알 수 없는 이벤트 타입: {event_type}")
        with self._db.transaction() as conn:
            row = conn.execute(
                "SELECT COALESCE(MAX(seq), 0) + 1 AS next FROM events WHERE project_id = ?",
                (project_id,),
            ).fetchone()
            event_id = f"evt_{uuid.uuid4().hex[:16]}"
            conn.execute(
                "INSERT INTO events (id, project_id, run_id, seq, type, payload_json, created_at)"
                " VALUES (?, ?, ?, ?, ?, ?, ?)",
                (event_id, project_id, run_id, row["next"], event_type,
                 json.dumps(payload, ensure_ascii=False), now_iso()),
            )
            stored = conn.execute("SELECT * FROM events WHERE id = ?", (event_id,)).fetchone()
        event = _to_event(stored)
        self._publish(project_id, event)
        return event

    def list_after(self, project_id: str, after_seq: int = 0) -> list[dict]:
        rows = self._db.query(
            "SELECT * FROM events WHERE project_id = ? AND seq > ? ORDER BY seq", (project_id, after_seq)
        )
        return [_to_event(row) for row in rows]

    def last_seq(self, project_id: str) -> int:
        row = self._db.one("SELECT COALESCE(MAX(seq), 0) AS seq FROM events WHERE project_id = ?", (project_id,))
        return row["seq"]

    # --- 실시간 전달 -------------------------------------------------------------
    # 실행 감시는 별도 스레드에서 돌기 때문에, 구독자의 이벤트 루프로 안전하게 넘긴다.

    def subscribe(self, project_id: str) -> _Subscriber:
        subscriber = _Subscriber(asyncio.get_running_loop())
        with self._sub_lock:
            self._subscribers.setdefault(project_id, set()).add(subscriber)
        return subscriber

    def unsubscribe(self, project_id: str, subscriber: _Subscriber) -> None:
        with self._sub_lock:
            self._subscribers.get(project_id, set()).discard(subscriber)

    def _publish(self, project_id: str, event: dict) -> None:
        with self._sub_lock:
            subscribers = list(self._subscribers.get(project_id, ()))
        for subscriber in subscribers:
            try:
                subscriber.loop.call_soon_threadsafe(subscriber.queue.put_nowait, event)
            except RuntimeError:  # 루프가 이미 닫힘(연결 종료)
                self.unsubscribe(project_id, subscriber)
